### PR TITLE
[lld][COFF][LTO] Implement /lldemit:llvm option

### PR DESCRIPTION
**Pull request by mizvekov**
With this new option, bitcode will be emited instead of object code. This is analogous to the `--plugin-opt=emit-llvm` option in the ELF linker.

### CODE REVIEW

File: lld/COFF/Driver.cpp [hunk 1]
```diff
@@ -2101,7 +2103,7 @@ void LinkerDriver::linkerMain(ArrayRef<const char *> argsArr) {
   // Handle /RELEASE
   if (args.hasArg(OPT_release))
     config->writeCheckSum = true;
-  
+
```
Comment (mstorsjo) [inline lld/COFF/Driver.cpp:2106]: Nit: Stray change

File: lld/Common/Filesystem.cpp [hunk 2]
```diff
@@ -57,7 +58,7 @@ void lld::unlinkAsync(StringRef path) {
   //
   // The code here allows LLD to work on all versions of Windows.
   // However, at Windows 10 1903 it seems that the behavior of
-  // Windows has changed, so that we could simply delete the output 
+  // Windows has changed, so that we could simply delete the output
```
Comment (mstorsjo) [inline lld/Common/Filesystem.cpp:61]: Nit: Stray change

(The cleanup of trailing whitespace certainly is good though, but it'd be better to keep it in a separate commit, which I guess you could push just bypassing review.)

File: lld/include/lld/Common/Filesystem.h [hunk 3]
```diff
@@ -10,11 +10,15 @@
...
 
 #include "lld/Common/LLVM.h"
+#include "llvm/Support/raw_ostream.h"
+#include <memory>
 #include <system_error>
 
 namespace lld {
 void unlinkAsync(StringRef path);
 std::error_code tryCreateFile(StringRef path);
+std::unique_ptr<llvm::raw_fd_ostream> openFile(StringRef file);
```
Comment (mstorsjo) [inline lld/include/lld/Common/Filesystem.h:20]: Could we have the refactoring of `openFile` and `openLTOOutputFile` in a separate preceding commit?

File: lld/COFF/Driver.cpp [hunk 4]
```diff
@@ -2393,7 +2395,8 @@ void LinkerDriver::linkerMain(ArrayRef<const char *> argsArr) {
   // If -thinlto-index-only is given, we should create only "index
   // files" and not object files. Index file creation is already done
   // in addCombinedLTOObject, so we are done if that's the case.
-  if (config->thinLTOIndexOnly)
+  // Likewise, for /emitllvm we only emit bitcode.
```
Comment (mstorsjo) [inline lld/COFF/Driver.cpp:2398]: The option isn't `/emitllvm` but `/opt:emitllvm`.

File: lld/COFF/Driver.cpp [hunk 5]
```diff
@@ -1822,6 +1822,8 @@ void LinkerDriver::linkerMain(ArrayRef<const char *> argsArr) {
         ltoDebugPM = true;
       } else if (s == "noltodebugpassmanager") {
         ltoDebugPM = false;
+      } else if (s == "emitllvm") {
```
Comment (mstorsjo) [inline lld/COFF/Driver.cpp:1825]: I'd defer to @rnk or someone else about whether `/opt:emitllvm` is the most natural/idiomatic way of setting an option like this on the `link.exe` like interface? AFAIK the `/opt:` namespace is for configuring optimizations, right? We have a bunch of other lld-link specific options either named plain `/customoption` or `/lldcustomoption`.

Secondly, while working on this, would you consider adding `emit-asm` in some form as well?

Comment (mizvekov) [reply lld/COFF/Driver.cpp:1826]: I have implemented emit-asm in https://github.com/llvm/llvm-project/pull/67079 .

Comment (mstorsjo) [reply lld/COFF/Driver.cpp:1826]: > I thought about this, but the best argument I have got so far is that this matches the equivalent option for the ELF linker, where it's in the `--plugin-opt=` umbrella.

Well, `--plugin-opt=` means "plugin options". But `/opt:` in (lld-)link means "optimization options", so I don't think it's the best fit. Or can @rnk chime in here if I'm misunderstanding things?

**Comment by mstorsjo** [comment]
I think the first commit looks good to go at this point; do you want to make a separate PR for that, that we can approve right away?

**Comment by aganea** [comment]
I lean on to what Martin says, I am not super excited about re-using `/opt:` for this purpose here (I do find the PR useful otherwise). Maybe `/lldemitasm`, `/lldemitllvm` unless someone has better naming?

**Comment by mizvekov** [comment]
> I lean on to what Martin says, I am not super excited about re-using `/opt:` for this purpose here (I do find the PR useful otherwise). Maybe `/lldemitasm`, `/lldemitllvm` unless someone has better naming?

I sure don't mind changing the flag to whatever is the best fit. I can wait for a second opinion from @rnk , and we can debate about the spelling all we want :)

**Comment by mizvekov** [comment]
I think `/lldemitasm` is good, in that it's less risky of future collisions with `link.exe`, but at the same time I find it hard on the eyes.
I wish we grouped all those `lld-link` exclusive options with a separator similar to `/opt`, something like `/lld:emitllvm` is nicer to read.

Another option is to have an `/outtype` flag that would default to `obj`, but also accept `llvm` and `asm`.
But that is still an `lld-link` exclusive option, and `/lldouttype` is also hard on the eyes.

**Comment by tru** [comment]
maybe `/lldemit:asm` `/lldemit:llvm`?

**Comment by mizvekov** [comment]
I have updated the MR (and subsequent emitasm one) to use a standalone flag.

**Review by mstorsjo — COMMENTED** [review]
I think this looks good overall, one minor nit. I'd love to hear more acks about the option naming from others like @aganea or @rnk here, but if nobody else wants to speak up, I'd be ok to approve this.

**Comment by aganea** [comment]
> I think this looks good overall, one minor nit. I'd love to hear more acks about the option naming from others like @aganea or @rnk here, but if nobody else wants to speak up, I'd be ok to approve this.

Naming SGTM.

**Review by rnk — APPROVED** [review]
I like the new name, thanks for the update.

**Comment by gregrodgers** [comment]
My build is failing with this change. 
FAILED: lib/liblldCOFF.so.18git
: && /usr/bin/c++ -fPIC -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-maybe-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wno-misleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG  -Wl,-z,defs -Wl,-z,nodelete   -Wl,-rpath-link,/home/grodgers/git/trunk18.0/build/llvm-project/./lib  -Wl,--gc-sections -shared -Wl,-soname,liblldCOFF.so.18git -o lib/liblldCOFF.so.18git tools/lld/COFF/CMakeFiles/lldCOFF.dir/CallGraphSort.cpp.o tools/lld/COFF/CMakeFiles/lldCOFF.dir/Chunks.cpp.o tools/lld/COFF/CMakeFiles/lldCOFF.dir/COFFLinkerContext.cpp.o tools/lld/COFF/CMakeFiles/lldCOFF.dir/DebugTypes.cpp.o tools/lld/COFF/CMakeFiles/lldCOFF.dir/DLL.cpp.o tools/lld/COFF/CMakeFiles/lldCOFF.dir/Driver.cpp.o tools/lld/COFF/CMakeFiles/lldCOFF.dir/DriverUtils.cpp.o tools/lld/COFF/CMakeFiles/lldCOFF.dir/ICF.cpp.o tools/lld/COFF/CMakeFiles/lldCOFF.dir/InputFiles.cpp.o tools/lld/COFF/CMakeFiles/lldCOFF.dir/LLDMapFile.cpp.o tools/lld/COFF/CMakeFiles/lldCOFF.dir/LTO.cpp.o tools/lld/COFF/CMakeFiles/lldCOFF.dir/MapFile.cpp.o tools/lld/COFF/CMakeFiles/lldCOFF.dir/MarkLive.cpp.o tools/lld/COFF/CMakeFiles/lldCOFF.dir/MinGW.cpp.o tools/lld/COFF/CMakeFiles/lldCOFF.dir/PDB.cpp.o tools/lld/COFF/CMakeFiles/lldCOFF.dir/SymbolTable.cpp.o tools/lld/COFF/CMakeFiles/lldCOFF.dir/Symbols.cpp.o tools/lld/COFF/CMakeFiles/lldCOFF.dir/Writer.cpp.o  -Wl,-rpath,"\$ORIGIN/../lib:/home/grodgers/git/trunk18.0/build/llvm-project/lib:"  lib/liblldCommon.so.18git  -lpthread  lib/libLLVMX86CodeGen.so.18git  lib/libLLVMX86AsmParser.so.18git  lib/libLLVMX86Desc.so.18git  lib/libLLVMX86Disassembler.so.18git  lib/libLLVMX86Info.so.18git  lib/libLLVMAMDGPUCodeGen.so.18git  lib/libLLVMAMDGPUAsmParser.so.18git  lib/libLLVMAMDGPUDisassembler.so.18git  lib/libLLVMLibDriver.so.18git  lib/libLLVMLTO.so.18git  lib/libLLVMPasses.so.18git  lib/libLLVMWindowsDriver.so.18git  lib/libLLVMWindowsManifest.so.18git  lib/libLLVMAMDGPUDesc.so.18git  lib/libLLVMAMDGPUInfo.so.18git  lib/libLLVMAMDGPUUtils.so.18git  lib/libLLVMDebugInfoDWARF.so.18git  lib/libLLVMDebugInfoPDB.so.18git  lib/libLLVMDebugInfoMSF.so.18git  lib/libLLVMObject.so.18git  lib/libLLVMMC.so.18git  lib/libLLVMDebugInfoCodeView.so.18git  lib/libLLVMCore.so.18git  lib/libLLVMBinaryFormat.so.18git  lib/libLLVMOption.so.18git  lib/libLLVMTargetParser.so.18git  lib/libLLVMSupport.so.18git  lib/libLLVMDemangle.so.18git  -Wl,-rpath-link,/home/grodgers/git/trunk18.0/build/llvm-project/lib && :
/usr/bin/ld: tools/lld/COFF/CMakeFiles/lldCOFF.dir/LTO.cpp.o: in function `std::_Function_handler<bool (unsigned int, llvm::Module const&), lld::coff::BitcodeCompiler::createConfig()::{lambda(unsigned long, llvm::Module const&)#1}>::_M_invoke(std::_Any_data const&, unsigned int&&, llvm::Module const&)':
LTO.cpp:(.text._ZNSt17_Function_handlerIFbjRKN4llvm6ModuleEEZN3lld4coff15BitcodeCompiler12createConfigEvEUlmS3_E_E9_M_invokeERKSt9_Any_dataOjS3_+0x50): undefined reference to `llvm::WriteBitcodeToFile(llvm::Module const&, llvm::raw_ostream&, bool, llvm::ModuleSummaryIndex const*, bool, std::array<unsigned int, 5ul>*)'
collect2: error: ld returned 1 exit status

**Comment by mizvekov** [comment]
Can you confirm this patch would help you?
```
diff --git a/lld/COFF/CMakeLists.txt b/lld/COFF/CMakeLists.txt
index acbd2e5fe2a5..2bbadf75bfa1 100644
--- a/lld/COFF/CMakeLists.txt
+++ b/lld/COFF/CMakeLists.txt
@@ -25,6 +25,7 @@ add_lld_library(lldCOFF
   LINK_COMPONENTS
   ${LLVM_TARGETS_TO_BUILD}
   BinaryFormat
+  BitWriter
   Core
   DebugInfoCodeView
   DebugInfoDWARF
```

**Comment by gregrodgers** [comment]
Yes, the above cmake fix corrected my build fail.  I look forward to your merge of the above cmake fix.  Thank you.

**Comment by mizvekov** [comment]
It's done, thanks!
